Collectors.py: Add support to ignore directories

mrtes opened the pull request:
This change allows users to specify directory globs in the form:
.../test/
.../test
.../test/**
.../test/*

These patterns now ignore the content of the whole directory and all
subdirectories.

Closes https://github.com/coala/coala/issues/4673

<!--
Thanks for your contribution!

Please take a quick look at those things down there. They're quite important.
Really! We wrote them for you. Yes you! With utmost care. Read them.
-->

**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [x] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [x] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `cobot mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

Asalle commented:
ack ae9a04cf8907cbc6ef2068bae6885bff728741bc

jayvdb commented on coalib/parsing/Globbing.py:
```diff
@@ -216,8 +217,8 @@ def fnmatch(name, globs):
 
-    if len(globs) == 0:
-        return True
+    if not globs:
```
Is this necessary?  It changes the return semantics of this function for edge cases.
best avoided, as it might break coala-quickstart or some other user of these functions which are not being tested by CI.

mrtes replied:
No its not necessary -> I will remove it.
Is there a list, which modules are used by other applications?
That would make it clearer, which functions can be modified safely.

jayvdb commented on coalib/collecting/Collectors.py:
```diff
@@ -54,11 +57,57 @@ def icollect(file_paths, ignored_globs=None, match_cache={}):
+    """
+    a, b = itertools.tee(((predicate(item), item) for item in iterable))
+
+    return ([item for pred, item in a if pred],
```
This runs the predicate twice for each item; when done for 1000s of files, that is significant.

mrtes replied:
I didn't test this, but the predicate should only be evaluated in the generator expression supplied to `itertools.tee` -> once per item (I guess I could remove one set of brackets there?).

In the return statement, only the _result_ of the predicate (i.e. a boolean) is compared, which should be fast.

There could be an  issue with large input iterables, that contain mostly _truthy_ or _falsy_ values (in the context of our predicate), see [the documentation](https://docs.python.org/2/library/itertools.html#itertools.tee).

jayvdb commented on coalib/collecting/Collectors.py:
```diff
@@ -54,11 +57,57 @@ def icollect(file_paths, ignored_globs=None, match_cache={}):
+        [])
+
+
+def collect_files(file_paths, log_printer=None, ignored_file_paths=(),
```
`ignored_file_paths=()` changes the semantics of the function, which is why you had to change some of the existing tests.

It is generally a bad idea to rely on args default being a empty collection.  The default should be None, and inside the function it does the right thing for any False-ish value.

i.e., inside, add `if not foo: foo = (, )` , as this is an internal matter.

As a general rule, a function should never distinguish between different types of False-ish provided in data given in arguments unless it was part of a very clear contract.

mrtes replied:
Can you explain why this is generally bad?
I understand that argument in case of empty list defaults, because the list is only allocated once and all changes values will carry over to the next function call.
However, this is not the case with tuples.

The advantage of `None` is that we are forced to check for Falsy-Values.
However, we also accept all other Falsy values, which can lead to weird behaviour, like accepting `0` and throwing when another number is provided.

In my opinion, using a tuple default value if a collection is expected makes the contract a lot clearer for people using the function.

jayvdb commented on tests/collecting/CollectorsTest.py:
```diff
@@ -35,7 +36,6 @@ def test_file_invalid(self):
         with LogCapture() as capture:
             self.assertEqual(collect_files(file_paths=['invalid_path'],
                                            log_printer=self.log_printer,
-                                           ignored_file_paths=None,
```
undo this please.

jayvdb commented on coalib/collecting/Collectors.py:
```diff
@@ -90,7 +147,7 @@ def collect_files(file_paths, log_printer=None, ignored_file_paths=None,
 
 
-def collect_dirs(dir_paths, ignored_dir_paths=None):
+def collect_dirs(dir_paths, ignored_dir_paths=()):
```
as above; empty list defaults are bad.

jayvdb commented:
unack https://github.com/coala/coala/commit/ae9a04cf8907cbc6ef2068bae6885bff728741bc

mrtes commented:
@jayvdb  I changed my PR according to your comments (with the exception of the partition function) and restructured the code a bit.

jayvdb commented on coalib/parsing/Globbing.py:
```diff
@@ -196,7 +196,8 @@ def translate(pattern):
 def fnmatch(name, globs):
     """
-    Tests whether name matches one of the given globs.
+    Tests whether name matches one of the given globs. An empty glob will
```
would be nice to add your addition starting on a new line, so the old line isnt modified.

jayvdb commented on tests/collecting/CollectorsTest.py:
```diff
@@ -94,10 +120,10 @@ def test_limited(self):
-                                                'others',
-                                                '*',
-                                                '*2.py')]),
+                              self.collectors_test_dir,
```
this change isnt necessary.
please undo

jayvdb commented on tests/collecting/CollectorsTest.py:
```diff
@@ -339,7 +365,8 @@ def test_filter_section_bears_by_languages(self):
                          "<class 'bears2.Test2LocalBear'>")
 
-        global_bears = filter_section_bears_by_languages(global_bears, ['Java'])
+        global_bears = filter_section_bears_by_languages(
```
this change wasnt necessary; please undo

jayvdb commented on coalib/processes/Processing.py:
```diff
@@ -382,7 +382,6 @@ def instantiate_processes(section,
     """
     filename_list = collect_files(
         glob_list(section.get('files', '')),
-        None,
```
you can drop this.  A newcomer will do it.
https://github.com/coala/coala/issues/5341

mrtes commented:
Implemented your suggestions.
Btw. the partition function I added, could be moved to some utils module, so that it can be used in other contexts, without requiring a dependency to the collectors module.
I'm just not sure which module would be suited for this.
Any suggestions?

jayvdb commented:
`coalib.misc.IterUtilities`?

or you can simply leave it and prefix it with `_`, and we worry about its final resting home another day.

mrtes commented:
Alright.
Going with your first suggestion :)

mrtes commented:
Not sure what the problems of circle ci and gitmate are.
My local tests succeed, and I haven't really changed much, except for moving the `partition` function.
Anyway the circle ci failures do not seem to be related to failing tests?

jayvdb commented:
@gitmate-bot rebase

jayvdb commented:
ack https://github.com/coala/coala/pull/4842/commits/3d8c8189e09dcd228752200dfdfefd9212e4dc7d

jayvdb commented:
@gitmate-bot ff
